File: Credicrew/backend/app/services/outreach.py
"""Deterministic outreach email composer.

Mirror of the frontend at `frontend/src/lib/outreach.ts`. Same templating
rules so the API and the in-browser modal produce identical drafts.

No LLM call; pure-string templating from the role's parsed plan, the JD
pitch line, and the candidate's match result.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable


@dataclass
class OutreachEmail:
    subject: str
    body: str


def first_name(full: str | None) -> str:
    if not full:
        return "there"
    return full.strip().split()[0] or "there"


def join_list(items: Iterable[str]) -> str:
    parts = [s for s in items if s]
    if not parts:
        return ""
    if len(parts) == 1:
        return parts[0]
    if len(parts) == 2:
        return f"{parts[0]} and {parts[1]}"
    return f"{', '.join(parts[:-1])}, and {parts[-1]}"


def title_case(s: str | None) -> str:
    return " ".join(w[:1].upper() + w[1:] for w in (s or "").split())


def extract_pitch(jd: str | None) -> str:
    """Pull the first sentence of the JD as a short pitch line."""
    if not jd:
        return ""
    trimmed = jd.strip()
    # First sentence terminator
    for term in (". ", ".\n", "! ", "!\n", "? ", "?\n"):
        idx = trimmed.find(term)
        if 0 < idx < 220:
            return trimmed[: idx + 1].strip()
    head = trimmed.split("\n", 1)[0]
    if len(head) > 220:
        return head[:217].rstrip() + "…"
    return head


def compose_email(
    *,
    role_name: str | None,
    plan_skills: list[str] | None,
    plan_location: str | None,
    plan_seniority: str | None,
    pitch: str | None,
    candidate_name: str | None,
    candidate_role: str | None,
    matched_skills: list[str] | None = None,
    score: int | None = None,
    sender: str | None = None,
) -> OutreachEmail:
    fn = first_name(candidate_name)
    role_title = role_name or (
        title_case(plan_seniority or "engineering") + " role"
    )
    matched = (matched_skills or [])[:3]
    pitch_line = (
        (pitch or "").strip()
        or f"we're hiring for {role_title.lower()} and your background looks like a strong fit"
    )

    if matched:
        subject = (
            f"Quick chat about a {role_title} role — your {matched[0]} work caught my eye"
        )
    else:
        subject = f"Quick chat about a {role_title} role at our team"

    if matched:
        skill_line = (
            f"Your work with {join_list(matched)} stood out — those are exactly "
            "the skills we need on day one."
        )
    else:
        skill_line = (
            f"Your background as a {candidate_role or 'engineer'} stood out across "
            "the candidates I reviewed."
        )

    if plan_location and plan_location != "remote":
        loc_line = (
            f"We're based in {title_case(plan_location)}, but happy to discuss "
            "remote / hybrid arrangements."
        )
    else:
        loc_line = "The role is remote-friendly across India."

    score_line = (
        f"(For context: against the spec for {role_title}, your profile lands at "
        f"{score}/100 with explainable reasons — happy to share the breakdown.)"
        if isinstance(score, int)
        else ""
    )

    lines = [
        f"Hi {fn},",
        "",
        f"I'm reaching out about a {role_title} opportunity — {pitch_line}",
        "",
        skill_line,
        loc_line,
        "",
        "Would you be open to a 20-minute intro call this week? Even if the timing "
        "isn't right, I'd love to keep in touch.",
        score_line,
        "",
        "Best,",
        sender or "— Hiring team",
    ]
    body = "\n".join(line for line in lines if line is not None)
    return OutreachEmail(subject=subject, body=body)

File: Credicrew/frontend/src/lib/outreach.ts
// Deterministic outreach email composer.
//
// No LLM call — pure-string templating from the candidate, the role's
// query plan, and the JD pitch line. Mirrored on the backend at
// app/services/outreach.py so the same email comes back from POST
// /outreach for programmatic / agentic clients.

import type { Candidate } from '@/data/candidates';
import type { MatchResult } from '@/lib/match';
import type { Role } from '@/lib/roles';

export type OutreachEmail = {
  subject: string;
  body: string;
};

function firstName(full: string): string {
  return (full || '').trim().split(/\s+/)[0] || 'there';
}

function joinList(items: string[]): string {
  if (items.length <= 1) return items[0] ?? '';
  if (items.length === 2) return `${items[0]} and ${items[1]}`;
  return `${items.slice(0, -1).join(', ')}, and ${items[items.length - 1]}`;
}

function titleCase(s: string): string {
  return s
    .split(/\s+/)
    .map(w => (w ? w[0].toUpperCase() + w.slice(1) : ''))
    .join(' ');
}

export type ComposeArgs = {
  role: Pick<Role, 'name' | 'plan' | 'pitch'>;
  candidate: Pick<Candidate, 'name' | 'role' | 'location'>;
  match?: MatchResult;
  sender?: string;
};

export function composeEmail(args: ComposeArgs): OutreachEmail {
  const { role, candidate, match, sender } = args;

  const fn = firstName(candidate.name);
  const roleTitle = role.name || titleCase(role.plan.seniority || 'engineering') + ' role';
  const matched = (match?.matchedSkills ?? []).slice(0, 3);
  const pitch =
    (role.pitch && role.pitch.trim()) ||
    `we're hiring for ${roleTitle.toLowerCase()} and your background looks like a strong fit`;

  const subject = matched.length
    ? `Quick chat about a ${roleTitle} role — your ${matched[0]} work caught my eye`
    : `Quick chat about a ${roleTitle} role at our team`;

  const skillLine = matched.length
    ? `Your work with ${joinList(matched)} stood out — those are exactly the skills we need on day one.`
    : `Your background as a ${candidate.role || 'engineer'} stood out across the candidates I reviewed.`;

  const locLine =
    role.plan.location && role.plan.location !== 'remote'
      ? `We're based in ${titleCase(role.plan.location)}, but happy to discuss remote / hybrid arrangements.`
      : `The role is remote-friendly across India.`;

  const score = match?.score;
  const scoreLine =
    typeof score === 'number'
      ? `(For context: against the spec for ${roleTitle}, your profile lands at ${score}/100 with explainable reasons — happy to share the breakdown.)`
      : '';

  const body = [
    `Hi ${fn},`,
    '',
    `I'm reaching out about a ${roleTitle} opportunity — ${pitch}`,
    '',
    skillLine,
    locLine,
    '',
    `Would you be open to a 20-minute intro call this week? Even if the timing isn't right, I'd love to keep in touch.`,
    scoreLine,
    '',
    `Best,`,
    sender || `— Hiring team`,
  ]
    .filter(line => line !== null && line !== undefined)
    .join('\n');

  return { subject, body };
}

export function toMailto(to: string | undefined, email: OutreachEmail): string {
  const params = new URLSearchParams();
  params.set('subject', email.subject);
  params.set('body', email.body);
  return `mailto:${to ?? ''}?${params.toString()}`;
}
